dialects/refly/refly.thorin

```diff
@@ -0,0 +1,47 @@
+/// # The refly dialect {#refly}
+///
+/// The refly dialect allows to [reify](https://en.wikipedia.org/wiki/Reification_(computer_science)) and [reflect](https://en.wikipedia.org/wiki/Reflective_programming) Thorin's own representation.
+///
+/// [TOC]
+///
+/// ## Dependencies
+///
+/// none
+///
+/// ## Types
+///
+.ax %refly.Code: *;
+///
+/// ## Reify/Reflect
+///
+/// ### %refly.reify
+///
+/// Yields the internal representation of a Thorin expression as `%%refly.Code`.
+.ax %refly.reify: Π T: * -> T -> %refly.Code, normalize_reify;
+///
+/// ### %refly.reflect
+///
+/// Converts a `%%refly.Code` back to a Thorin expression.
+.ax %refly.reflect: Π T: * -> %refly.Code -> T, normalize_reflect;
+///
+/// ## Inspect
+///
+/// ### %refly.gid
+///
+/// Retuns the internal thorin::Def::gid of the argument.
+.ax %refly.gid: Π T: * -> T -> .Nat, normalize_gid;
+///
+/// ### %refly.dbg
+/// 
+/// Debugs a given expression.
+/// * `tmp`: Prints debug information of a given expression at construction time and vanishes afterwrads.
+/// * `perm`: Permanently debugs at every construction.
+///   Gets removed at codegen preparation in thorin::refly::RemoveDbgPerm.
+.ax %refly.dbg(tmp, perm): Π T: * -> T -> T, normalize_dbg;
+/// 
+/// ## Manipulate
+/// 
+/// ### %refly.refine
+///
+/// Sets the `i`th operand of the reified Code `e` to `x`.
+.ax %refly.refine: Π T: * -> [e: %refly.Code, i: .Nat, x: T] -> %refly.Code, normalize_refine;
```

Comment:
An alternative approach (with larger meta programming capabilities in mind) would be a `getOps` and `rebuild (exp, ops)` axiom.
It might also be strange to insert real code into quoted terms instead of working the quoted world with quoted terms.

Reply:
I just wanted to have one simple operation that allows us to manipulate code. If we are picking up on this - sure - we basically need to rebuild the important parts of the C++-API with `get_ops`, `rebuild`, `stub`, etc. But I think for now, the `%refly.refine` operation is a pretty simple example to manipulate sth.

Reply:
Initially I thought, that it's cool to showcase how you can mix reified and actual expressions, but yes, that is too much. I changed the signature accordingly.